test/cpp/api/dataloader.cpp

```diff
@@ -2224,4 +2224,78 @@ TEST(DataLoaderTest, ChunkDatasetCrossChunkShuffle) {
           std::equal(result.begin(), result.end(), expected_result.begin()));
     }
   }
+}
+
+TEST(DataLoaderTest, CustomSortPolicy) {
+  const size_t chunk_size = 5;
+  const size_t batch_size = 10;
+
+  struct D : public datasets::ChunkDataReader<int> {
+   public:
+    using BatchType = datasets::ChunkDataReader<int>::ChunkType;
+    D(size_t chunk_count) : chunk_count_(chunk_count) { }
+
+    BatchType read_chunk(size_t chunk_index) override {
+      BatchType batch_data(chunk_size);
+      auto rand_gen = [](){ return std::rand()%100; };
+      std::generate(batch_data.begin(), batch_data.end(), rand_gen);
+      return batch_data;
+    }
+
+    size_t chunk_count() override {
+      return chunk_count_;
+    };
+
+    void reset() override{};
+    size_t chunk_count_;
+  };
+
+  auto sorting_policy = [](std::vector<int>& raw_batch_data){
```

Comment:
Would `D::BatchType&` be more flexible than std::vector<int>, as it will still be back compatible if `D` implementation changes to another type in the future?

Reply:
Yes, it will be. I'll make the change in next iteration.